### ISSUE TITLE
MoE routing weights implementation mismatch

### ISSUE THREAD

**Issue by RaymondLi0** [ServiceNow/Fast-LLM#190]
I am not certain of the impact this could have, but just writing this down to keep track of this potential issue.

The routing-weight computation is not the same in Fast-LLM and in the Mixtral HF implementation:
- Fast-LLM: perform softmax on the top-k-logits https://github.com/ServiceNow/Fast-LLM/blob/main/fast_llm/layers/transformer/mixture_of_experts.py#L174
- HF Mixtral: softmax is applied on all the logits, and then the top-k probs are normalized https://github.com/huggingface/transformers/blob/main/src/transformers/models/mixtral/modeling_mixtral.py#L122-L124
- In Switch-transformers https://arxiv.org/pdf/2101.03961 (sec 2.1), they describe yet another procedure: softmax is applied on all the logits, and the un-normalized top-k probs are used

Note that in the Mixtral report https://arxiv.org/pdf/2401.04088 (sec 2.1) they describe what we do in Fast-LLM, not what's in the HF implementation!

**Comment by RaymondLi0** [comment]
Nevermind, the first two items result in the same weights in the end
